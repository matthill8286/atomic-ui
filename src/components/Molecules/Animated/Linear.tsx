import React from 'react'
import { animated, useSpring } from 'react-spring'
import { styled } from '@/styles/styled'

interface LineProps {
  lineWidth?: number
  lineHeight?: number
  lineMargin?: number
  backgroundColor?: string
  hoverColor: string
}

const StyledOuterLine = styled.div`
  cursor: pointer;
  border-radius: ${({ lineHeight }: LineProps) => (lineHeight ? `${lineHeight / 2}px` : '4px')};
  overflow: hidden;
  width: ${({ lineWidth }: LineProps) => (lineWidth ? `${lineWidth}px` : '50px')};
  margin: ${({ lineMargin }: LineProps) => lineMargin && lineMargin};
  height: ${({ lineHeight }: LineProps) => (lineHeight ? `${lineHeight}px` : '8px')};
  background-color: ${({ backgroundColor }: LineProps) =>
    backgroundColor ? backgroundColor : 'grey'};

  &:hover {
    background-color: ${({ hoverColor }: LineProps) => hoverColor};
  }
`

const StyledInnerLine = styled.div`
  height: ${({ lineHeight }: LineProps) => (lineHeight ? `${lineHeight}px` : '8px')};
  background-color: ${({ backgroundColor }: LineProps) => backgroundColor ?? 'black'};
`

const AnimatedOuterLine = animated(StyledOuterLine)
const AnimatedInnerLine = animated(StyledInnerLine)

interface DotsProps {
  index: number
  current?: number
  duration?: number
  width?: number
  lineStroke?: number
  selected?: boolean
  color?: string
  backgroundColor?: string
  margin?: string
  onClick: ({ index }: { index: number }) => void
}

export const Linear = ({
  index,
  current = -1,
  duration = 4000,
  width = 35,
  lineStroke = 1,
  selected = false,
  color = '#918e8c',
  backgroundColor = '#cfcbca',
  margin,
  onClick,
}: DotsProps) => {
  const active = selected || index === current

  const { lineWidth } = useSpring({
    config: { duration },
    from: { lineWidth: 0 },
    lineWidth: 100,
    reset: active,
  })

  return (
    <AnimatedOuterLine
      hoverColor={color}
      onClick={onClick}
      lineWidth={width}
      lineHeight={lineStroke}
      backgroundColor={backgroundColor}
      lineMargin={margin}>
      {active && (
        <AnimatedInnerLine
          style={{ width: lineWidth.interpolate(x => `${x}%`) }}
          lineHeight={lineStroke}
          backgroundColor={color}
        />
      )}
    </AnimatedOuterLine>
  )
}
