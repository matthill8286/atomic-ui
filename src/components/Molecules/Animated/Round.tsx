import React from 'react'
import { animated, config, useSpring } from 'react-spring'
import { css, styled } from '@/styles/styled'

const StyledCircle = styled.circle`
  -moz-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  -o-transform-origin: 50% 50%;
  -webkit-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  cursor: pointer;
`

interface CircleProps {
  hoverColor: string
}

const StyledUnactive = styled(StyledCircle)`
  &:hover {
    fill: ${({ hoverColor }: CircleProps) => hoverColor};
  }
`

interface IconProps {
  dotMargin?: string
}

const StyledIcon = styled.svg`
  ${({ dotMargin }: IconProps) =>
    dotMargin &&
    css`
      margin: ${dotMargin};
    `}
`

const AnimatedIcon = animated(StyledIcon)
const AnimatedCircle = animated(StyledCircle)

interface RoundProps {
  index: number
  current?: number
  duration?: number
  width?: number
  lineStroke?: number
  dotDifference?: number
  selected?: boolean
  color?: string
  margin?: string
  springLoading?: boolean
  onClick: ({ index }: { index: number }) => void
  withAnimation?: boolean
}

export const Round = ({
  index,
  current = -1,
  duration = 4000,
  width = 24,
  lineStroke = 1,
  dotDifference = 8,
  selected = false,
  color = 'grey',
  margin,
  springLoading = true,
  onClick,
  withAnimation = false,
}: RoundProps) => {
  const active = selected || index === current

  const dotWidth = width / 2 - dotDifference

  const fullRadius = width / 2
  const radius = width / 2 - lineStroke / 2
  const circumference = radius * 2 * Math.PI

  const { strokeOffset } = useSpring({
    config: { duration },
    from: { strokeOffset: 0 },
    strokeOffset: withAnimation ? 100 : 0,
    reset: active,
  })

  const { rotation } = useSpring({
    config: springLoading ? config.molasses : { duration },
    from: { rotation: -90 },
    rotation: withAnimation ? 270 : -90,
    reset: active,
  })

  return (
    <AnimatedIcon height={width} width={width} key={index} onClick={onClick} dotMargin={margin}>
      {active && (
        <AnimatedCircle
          r={radius}
          cx={fullRadius}
          cy={fullRadius}
          strokeWidth={lineStroke}
          stroke={color}
          fill={'transparent'}
          transform={rotation.interpolate(x => `rotate(${x})`)}
          strokeLinecap={'round'}
          strokeDashoffset={strokeOffset.interpolate(
            x => circumference - (x / 100) * circumference
          )}
          strokeDasharray={`${circumference} ${circumference}`}
        />
      )}
      <StyledUnactive
        hoverColor={color}
        r={dotWidth}
        fill={active ? color : 'transparent'}
        cx={fullRadius}
        cy={fullRadius}
        strokeWidth={lineStroke}
        stroke={color}
        strokeDasharray={0}
      />
    </AnimatedIcon>
  )
}
