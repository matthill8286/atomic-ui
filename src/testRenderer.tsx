import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
