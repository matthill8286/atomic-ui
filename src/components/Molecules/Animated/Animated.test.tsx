import React from 'react'
import { mountWithTheme, renderWithTheme } from '@/testRenderer'
import { Linear } from './Linear'
import { Round } from './Round'

describe('Pagination', () => {
  it('renders Linear Animated Pagination', () => {
    const options = {
      onClick: jest.fn(),
    }
    const tree = renderWithTheme(<Linear index={1} {...options} />)
    expect(tree).toMatchSnapshot()
  })
  it('renders Round Animated Pagination', () => {
    const options = {
      onClick: jest.fn(),
    }
    const tree = renderWithTheme(<Round index={1} {...options} />)
    expect(tree).toMatchSnapshot()
  })
  it('dot handles click', () => {
    const onClick = jest.fn()
    const options = {
      onClick,
    }
    const tree = mountWithTheme(<Round index={1} {...options} />)

    tree.simulate('click')
    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
