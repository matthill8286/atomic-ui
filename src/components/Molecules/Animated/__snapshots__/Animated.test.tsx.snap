// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders Linear Animated Pagination 1`] = `
.c0 {
  cursor: pointer;
  border-radius: 0.5px;
  overflow: hidden;
  width: 35px;
  height: 1px;
  background-color: #cfcbca;
}

.c0:hover {
  background-color: #918e8c;
}

<div
  className="c0"
  onClick={[MockFunction]}
/>
`;

exports[`Pagination renders Round Animated Pagination 1`] = `
.c0 {
  -moz-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  -o-transform-origin: 50% 50%;
  -webkit-transform-origin: 50% 50%;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  cursor: pointer;
}

.c1:hover {
  fill: grey;
}

<svg
  className=""
  height={24}
  onClick={[MockFunction]}
  width={24}
>
  <circle
    className="c0 c1"
    cx={12}
    cy={12}
    fill="transparent"
    r={4}
    stroke="grey"
    strokeDasharray={0}
    strokeWidth={1}
  />
</svg>
`;
